File: ios/EggTimerCore/Sources/EggTimerCore/Solve.swift
import Foundation

/// Putting it together: simulate a cook, and invert for the cook time that hits
/// a doneness target. Transliterated from `src/core/solve.ts`.

/// The calibratable parameters. Everything else is fixed physics.
public struct ModelParams: Sendable {
    /// Thermal diffusivity, m^2/s. Absorbs all geometry and property error,
    /// since only tau = R^2/alpha is identifiable.
    public var alphaM2s: Double
    /// Multiplier on the still-air cooling time constant.
    public var tauAirScale: Double

    public init(alphaM2s: Double = Constants.alphaDefault, tauAirScale: Double = 1.0) {
        self.alphaM2s = alphaM2s
        self.tauAirScale = tauAirScale
    }

    public static let `default` = ModelParams()
}

/// Dose target for the white, equivalent minutes at 80 C, read at the yolk
/// boundary - the innermost white is the last to set.
public let whiteDoseTarget = 0.05

/// Yolk dose targets at the ends of the slider, equivalent minutes at 63 C.
public let yolkDoseRunny = 0.05
public let yolkDoseHard = 2000.0

public struct Doneness: Sendable {
    public let level: Double
    public let yolkDoseMin: Double
    public let whiteDoseMin: Double
}

/// Map the single slider to a yolk dose target. Logarithmic, which - because
/// dose goes as 10^(T/z) - makes the slider linear in peak yolk temperature.
public func donenessFromSlider(_ level: Double) -> Doneness {
    let clamped = level < 0.0 ? 0.0 : (level > 1.0 ? 1.0 : level)
    let lo = log10(yolkDoseRunny)
    let hi = log10(yolkDoseHard)
    return Doneness(
        level: clamped,
        yolkDoseMin: pow(10.0, lo + (hi - lo) * clamped),
        whiteDoseMin: whiteDoseTarget
    )
}

public func sliderFromYolkDose(_ yolkDoseMin: Double) -> Double {
    let lo = log10(yolkDoseRunny)
    let hi = log10(yolkDoseHard)
    let level = (log10(yolkDoseMin) - lo) / (hi - lo)
    return level < 0.0 ? 0.0 : (level > 1.0 ? 1.0 : level)
}

public struct DonenessAnchor: Sendable {
    public let label: String
    public let level: Double
    public let approxPeakYolkC: Double
}

public let donenessAnchors: [DonenessAnchor] = [
    DonenessAnchor(label: "Runny", level: 0.00, approxPeakYolkC: 56),
    DonenessAnchor(label: "Soft", level: 0.22, approxPeakYolkC: 61),
    DonenessAnchor(label: "Jammy", level: 0.41, approxPeakYolkC: 65),
    DonenessAnchor(label: "Fudgy", level: 0.62, approxPeakYolkC: 70),
    DonenessAnchor(label: "Hard", level: 1.00, approxPeakYolkC: 77),
]

public struct CookResult: Sendable {
    public let cookTimeS: Double
    /// Highest temperature the yolk centre ever reaches, carryover included.
    public let peakYolkC: Double
    public let peakYolkTimeS: Double
    /// Yolk centre at the moment the egg leaves the water.
    public let yolkAtPullC: Double
    public let yolkDoseMin: Double
    public let whiteDoseMin: Double
    public let peakWhiteC: Double
}

/// Run one cook and report what it does to the egg.
public func simulate(
    egg: Egg, setup: CookSetup, params: ModelParams, cookTimeS: Double
) -> CookResult {
    let initialSurface = Protocols.initialSurfaceTemperature(setup)
    var sphere = SphereState(
        radiusM: egg.radiusM, alphaM2s: params.alphaM2s,
        initialC: setup.eggStartC, surfaceC: initialSurface
    )
    var yolkDose = Dose(zK: Constants.zYolk, trefC: Constants.tRefYolkC)
    var whiteDose = Dose(zK: Constants.zWhite, trefC: Constants.tRefWhiteC)

    var t = 0.0
    var peakYolk = setup.eggStartC
    var peakYolkTime = 0.0
    var peakWhite = setup.eggStartC
    var yolkAtPull = setup.eggStartC
    var pullRecorded = false
    var prevYolk = setup.eggStartC
    var peakDoseRate = 0.0
    // Captured when the egg leaves the water: a lumped egg in air relaxes from
    // its own volume-average temperature, the ceiling on carryover.
    var meanAtPull = setup.eggStartC
    var waterAtPull = initialSurface

    let endTime = cookTimeS + Constants.carryoverWindow
    while t < endTime {
        let tNext = t + Constants.dtSim
        let next: Double
        if tNext < cookTimeS {
            next = Protocols.bathTemperature(setup, tS: tNext)
        } else {
            if !pullRecorded {
                meanAtPull = sphere.meanTemperature
                waterAtPull = sphere.surfaceC
            }
            next = Protocols.coolingTemperature(
                setup: setup, elapsedSincePullS: tNext - cookTimeS,
                waterAtPullC: waterAtPull, meanAtPullC: meanAtPull,
                tauAirScale: params.tauAirScale
            )
        }
        sphere.step(dtS: Constants.dtSim, nextSurfaceC: next)
        t = tNext

        let yolkCentre = sphere.centreTemperature
        let whiteInner = sphere.temperature(atX: Constants.yolkRadiusFrac)
        yolkDose.accumulate(temperatureC: yolkCentre, dtS: Constants.dtSim)
        whiteDose.accumulate(temperatureC: whiteInner, dtS: Constants.dtSim)

        if yolkCentre > peakYolk {
            peakYolk = yolkCentre
            peakYolkTime = t
        }
        if whiteInner > peakWhite { peakWhite = whiteInner }
        if !pullRecorded && t >= cookTimeS {
            yolkAtPull = yolkCentre
            pullRecorded = true
        }

        // Stop once the egg is past its peak and the remaining dose rate is a
        // millionth of the peak: further integration cannot change the answer.
        let rate = pow(10.0, (yolkCentre - Constants.tRefYolkC) / Constants.zYolk)
        if rate > peakDoseRate { peakDoseRate = rate }
        if t > cookTimeS && yolkCentre < prevYolk && rate < 1e-6 * peakDoseRate { break }
        prevYolk = yolkCentre
    }

    return CookResult(
        cookTimeS: cookTimeS,
        peakYolkC: peakYolk,
        peakYolkTimeS: peakYolkTime,
        yolkAtPullC: yolkAtPull,
        yolkDoseMin: yolkDose.minutes,
        whiteDoseMin: whiteDose.minutes,
        peakWhiteC: peakWhite
    )
}

// MARK: - search

private let solveLoS = 20.0
private let solveHiS = 3600.0
private let solveTolS = 1.0

private func yolkOf(_ r: CookResult) -> Double { r.yolkDoseMin }
private func whiteOf(_ r: CookResult) -> Double { r.whiteDoseMin }

/// Cook time in [lo, hi] at which `metric` first reaches `target`, assuming the
/// metric is monotonic across that bracket. Returns the UPPER end of the final
/// bracket: the shortest cook known to meet the target. The midpoint would be
/// as accurate but could sit a hair short, and callers compare the dose at the
/// answer against the target.
private func bisectBetween(
    egg: Egg, setup: CookSetup, params: ModelParams,
    target: Double, metric: (CookResult) -> Double, loS: Double, hiS: Double
) -> Double {
    var lo = loS
    var hi = hiS
    while hi - lo > solveTolS {
        let mid = 0.5 * (lo + hi)
        if metric(simulate(egg: egg, setup: setup, params: params, cookTimeS: mid)) < target {
            lo = mid
        } else {
            hi = mid
        }
    }
    return hi
}

/// Bisect the whole range. Both doses increase monotonically with cook time
/// while the water is held at the boil, so this is exact there.
private func bisect(
    egg: Egg, setup: CookSetup, params: ModelParams,
    target: Double, metric: (CookResult) -> Double
) -> Double {
    bisectBetween(
        egg: egg, setup: setup, params: params, target: target, metric: metric,
        loS: solveLoS, hiS: solveHiS
    )
}

public struct Solution: Sendable {
    public let result: CookResult
    /// False when the requested doneness cannot be reached without leaving the
    /// white undercooked.
    public let reachable: Bool
    /// Shortest cook that still sets the white.
    public let minCookTimeS: Double
    /// Softest yolk doneness achievable, as a slider position.
    public let softestLevel: Double
    /// Hardest achievable. 1 whenever the water stays at the boil; less when
    /// the heat is off and the pan runs out before the yolk gets there.
    public let hardestLevel: Double
    /// False when the white never sets at all - only possible with the heat
    /// off, where the water can fall past the white's target while the egg is
    /// still in it.
    public let whiteSets: Bool
}

/// Solve for the cook time that delivers the requested doneness.
public func solveCookTime(
    egg: Egg, setup: CookSetup, params: ModelParams, doneness: Doneness
) -> Solution {
    if setup.afterBoil == .off {
        return solveStanding(egg: egg, setup: setup, params: params, doneness: doneness)
    }

    let minCook = bisect(
        egg: egg, setup: setup, params: params, target: doneness.whiteDoseMin, metric: whiteOf
    )
    let atMin = simulate(egg: egg, setup: setup, params: params, cookTimeS: minCook)
    let softestLevel = sliderFromYolkDose(atMin.yolkDoseMin)
    let whiteSets = atMin.whiteDoseMin >= doneness.whiteDoseMin

    if atMin.yolkDoseMin >= doneness.yolkDoseMin {
        // Even the shortest white-setting cook overcooks the yolk.
        return Solution(
            result: atMin, reachable: false, minCookTimeS: minCook,
            softestLevel: softestLevel, hardestLevel: 1.0, whiteSets: whiteSets
        )
    }

    let cook = bisect(
        egg: egg, setup: setup, params: params, target: doneness.yolkDoseMin, metric: yolkOf
    )
    return Solution(
        result: simulate(egg: egg, setup: setup, params: params, cookTimeS: cook),
        reachable: true, minCookTimeS: minCook,
        softestLevel: softestLevel, hardestLevel: 1.0, whiteSets: whiteSets
    )
}

// MARK: - standing

/// Fraction of the best available dose that counts as "as far as this pan
/// goes". The maximum sits on a plateau, so the time worth printing is the
/// start of it, not its peak.
private let standingKnee = 0.99

/// How long past the boil it is worth looking with the heat off.
private let standingHorizonS = 1800.0

/// Coarse step for the standing scan: fine enough that the bracket it hands to
/// the bisection is locally monotonic, coarse enough to stay cheap.
private let scanStepS = 30.0

/// Both doses sampled at every scan step. One pass answers every question the
/// standing solver has.
private struct DoseCurve {
    var timesS: [Double] = []
    var yolk: [Double] = []
    var white: [Double] = []
}

private func scanStanding(
    egg: Egg, setup: CookSetup, params: ModelParams, horizonS: Double
) -> DoseCurve {
    var curve = DoseCurve()
    var t = solveLoS
    while t <= horizonS {
        let r = simulate(egg: egg, setup: setup, params: params, cookTimeS: t)
        curve.timesS.append(t)
        curve.yolk.append(r.yolkDoseMin)
        curve.white.append(r.whiteDoseMin)
        t += scanStepS
    }
    return curve
}

private func indexOfMax(_ values: [Double]) -> Int {
    var best = 0
    for i in 1..<values.count where values[i] > values[best] { best = i }
    return best
}

/// First cook time at which a sampled dose reaches `target`, WITHOUT assuming
/// monotonicity - or -1 if it never does.
///
/// With the heat off, pulling later means pulling from cooler water, so the
/// carryover that follows is smaller and past a point the total dose FALLS with
/// a longer cook. Bisection on that does not lose accuracy, it lands anywhere.
/// So: walk the samples to the first crossing, then bisect inside that one
/// step, where the function is still rising.
private func firstCrossing(
    egg: Egg, setup: CookSetup, params: ModelParams,
    curve: DoseCurve, values: [Double], target: Double, metric: (CookResult) -> Double
) -> Double {
    for i in 0..<values.count {
        if values[i] < target { continue }
        if i == 0 { return curve.timesS[0] }
        return bisectBetween(
            egg: egg, setup: setup, params: params, target: target, metric: metric,
            loS: curve.timesS[i - 1], hiS: curve.timesS[i]
        )
    }
    return -1.0
}

/// The same question with the heat off, where neither dose is monotonic in cook
/// time and every shortcut above stops being valid. The scan runs to the
/// horizon regardless of where the target is crossed, so `hardestLevel` is the
/// true ceiling of this pan and not merely "at least what was asked".
private func solveStanding(
    egg: Egg, setup: CookSetup, params: ModelParams, doneness: Doneness
) -> Solution {
    let curve = scanStanding(
        egg: egg, setup: setup, params: params, horizonS: setup.timeToBoilS + standingHorizonS
    )
    let whiteCook = firstCrossing(
        egg: egg, setup: setup, params: params, curve: curve, values: curve.white,
        target: doneness.whiteDoseMin, metric: whiteOf
    )
    let yolkCook = firstCrossing(
        egg: egg, setup: setup, params: params, curve: curve, values: curve.yolk,
        target: doneness.yolkDoseMin, metric: yolkOf
    )

    let peak = indexOfMax(curve.yolk)
    let maxDose = curve.yolk[peak]
    let maxAt = curve.timesS[peak]
    let hardestLevel = sliderFromYolkDose(maxDose)
    let whiteSets = whiteCook > 0
    let minCook = whiteSets ? whiteCook : maxAt
    let atMin = simulate(egg: egg, setup: setup, params: params, cookTimeS: minCook)
    let softestLevel = sliderFromYolkDose(atMin.yolkDoseMin)

    if !whiteSets || yolkCook < 0 {
        // Either the water never gets the white where it needs to go, or it
        // runs out before the yolk does. Answer with the furthest this pan
        // goes - but the START of the plateau, since waiting past it achieves
        // nothing.
        let knee = firstCrossing(
            egg: egg, setup: setup, params: params, curve: curve, values: curve.yolk,
            target: standingKnee * maxDose, metric: yolkOf
        )
        let at = knee > 0 ? knee : maxAt
        return Solution(
            result: simulate(egg: egg, setup: setup, params: params, cookTimeS: at),
            reachable: false, minCookTimeS: minCook,
            softestLevel: softestLevel, hardestLevel: hardestLevel, whiteSets: whiteSets
        )
    }

    if atMin.yolkDoseMin >= doneness.yolkDoseMin {
        return Solution(
            result: atMin, reachable: false, minCookTimeS: minCook,
            softestLevel: softestLevel, hardestLevel: hardestLevel, whiteSets: whiteSets
        )
    }

    // Both constraints have to hold at the same pull: take the later crossing.
    let cook = yolkCook > whiteCook ? yolkCook : whiteCook
    return Solution(
        result: simulate(egg: egg, setup: setup, params: params, cookTimeS: cook),
        reachable: true, minCookTimeS: minCook,
        softestLevel: softestLevel, hardestLevel: hardestLevel, whiteSets: whiteSets
    )
}

File: src/core/solve.ts
/**
 * Putting it together: simulate a cook, and invert for the cook time that hits
 * a doneness target.
 */

import {
  ALPHA_DEFAULT, YOLK_RADIUS_FRAC, Z_YOLK, TREF_YOLK_C, Z_WHITE, TREF_WHITE_C,
  DT_SIM, CARRYOVER_WINDOW,
} from './constants.js';
import { Egg } from './geometry.js';
import {
  CookSetup, bathTemperature, coolingTemperature, initialSurfaceTemperature,
} from './protocol.js';
import {
  createSphere, stepSphere, temperatureAt, centreTemperature, meanTemperature,
} from './sphere.js';
import { Dose, createDose, accumulateDose } from './kinetics.js';

/** The calibratable parameters. Everything else is fixed physics. */
export interface ModelParams {
  /** Thermal diffusivity, m^2/s. Absorbs all geometry/property model error,
   *  since only tau = R^2/alpha is identifiable. */
  alpha_m2s: number;
  /** Multiplier on the still-air cooling time constant. Only identifiable if
   *  the user actually varies the cooling protocol. */
  tauAirScale: number;
}

export const DEFAULT_PARAMS: ModelParams = {
  alpha_m2s: ALPHA_DEFAULT,
  tauAirScale: 1.0,
};

/** Dose target for the white, equivalent minutes at 80 C. The binding white
 *  constraint is at the yolk boundary - the innermost white is the last to
 *  set, which is exactly the radius Williams' 0.76 encodes.
 *
 *  Calibrated so the shortest white-setting cook is ~5.8 min for a fridge-cold
 *  large egg dropped into boiling water, giving a peak inner-white temperature
 *  near 74 C. That is genuinely set but still tender: ovotransferrin (62 C) and
 *  lysozyme (~70 C) have denatured, ovalbumin (80 C) has not. Demanding full
 *  ovalbumin set instead would forbid the classic soft-boiled egg, which
 *  plainly exists. */
export const WHITE_DOSE_TARGET = 0.05;

/** Yolk dose targets at the ends of the slider, equivalent minutes at 63 C.
 *  Interpolated logarithmically, which - because dose goes as 10^(T/z) - is
 *  equivalent to interpolating PEAK YOLK TEMPERATURE linearly. So the slider is
 *  perceptually even, spanning roughly 56 C (runny) to 77 C (hard). */
export const YOLK_DOSE_RUNNY = 0.05;
export const YOLK_DOSE_HARD = 2000.0;

export interface Doneness {
  /** Slider position in [0, 1]: 0 = runny, 1 = hard. */
  level: number;
  yolkDose_min: number;
  whiteDose_min: number;
}

/** Map the single slider to a yolk dose target. */
export function donenessFromSlider(level: number): Doneness {
  const clamped = level < 0.0 ? 0.0 : level > 1.0 ? 1.0 : level;
  const lo = Math.log10(YOLK_DOSE_RUNNY);
  const hi = Math.log10(YOLK_DOSE_HARD);
  return {
    level: clamped,
    yolkDose_min: Math.pow(10.0, lo + (hi - lo) * clamped),
    whiteDose_min: WHITE_DOSE_TARGET,
  };
}

/** Inverse of the above - which slider position corresponds to a given dose. */
export function sliderFromYolkDose(yolkDose_min: number): number {
  const lo = Math.log10(YOLK_DOSE_RUNNY);
  const hi = Math.log10(YOLK_DOSE_HARD);
  const level = (Math.log10(yolkDose_min) - lo) / (hi - lo);
  return level < 0.0 ? 0.0 : level > 1.0 ? 1.0 : level;
}

/** Slider positions worth labelling, for the UI. Positions follow from the
 *  dose scale above; the temperatures are what the model actually predicts. */
export interface DonenessAnchor {
  label: string;
  level: number;
  approxPeakYolk_C: number;
}

export const DONENESS_ANCHORS: DonenessAnchor[] = [
  { label: 'Runny', level: 0.00, approxPeakYolk_C: 56 },
  { label: 'Soft', level: 0.22, approxPeakYolk_C: 61 },
  { label: 'Jammy', level: 0.41, approxPeakYolk_C: 65 },
  { label: 'Fudgy', level: 0.62, approxPeakYolk_C: 70 },
  { label: 'Hard', level: 1.00, approxPeakYolk_C: 77 },
];

export interface CookResult {
  cookTime_s: number;
  /** Highest temperature the yolk centre ever reaches, including carryover. */
  peakYolk_C: number;
  peakYolkTime_s: number;
  /** Yolk centre temperature at the moment the egg leaves the water. */
  yolkAtPull_C: number;
  yolkDose_min: number;
  whiteDose_min: number;
  peakWhite_C: number;
}

/** Run one cook and report what it does to the egg. */
export function simulate(
  egg: Egg, setup: CookSetup, params: ModelParams, cookTime_s: number,
): CookResult {
  const initialSurface = initialSurfaceTemperature(setup);
  const sphere = createSphere(egg.radius_m, params.alpha_m2s, setup.eggStart_C, initialSurface);
  const yolkDose: Dose = createDose(Z_YOLK, TREF_YOLK_C);
  const whiteDose: Dose = createDose(Z_WHITE, TREF_WHITE_C);

  let t = 0.0;
  let peakYolk = setup.eggStart_C;
  let peakYolkTime = 0.0;
  let peakWhite = setup.eggStart_C;
  let yolkAtPull = setup.eggStart_C;
  let pullRecorded = false;
  let prevYolk = setup.eggStart_C;
  let peakDoseRate = 0.0;
  // Captured when the egg leaves the water: a lumped egg in air relaxes from
  // its own volume-average temperature, which is also the ceiling on carryover.
  let meanAtPull = setup.eggStart_C;
  let waterAtPull = initialSurface;

  const endTime = cookTime_s + CARRYOVER_WINDOW;
  while (t < endTime) {
    const tNext = t + DT_SIM;
    let next: number;
    if (tNext < cookTime_s) {
      next = bathTemperature(setup, tNext);
    } else {
      if (!pullRecorded) {
        meanAtPull = meanTemperature(sphere);
        waterAtPull = sphere.surface_C;
      }
      next = coolingTemperature(
        setup, tNext - cookTime_s, waterAtPull, meanAtPull, params.tauAirScale,
      );
    }
    stepSphere(sphere, DT_SIM, next);
    t = tNext;

    const yolkCentre = centreTemperature(sphere);
    const whiteInner = temperatureAt(sphere, YOLK_RADIUS_FRAC);
    accumulateDose(yolkDose, yolkCentre, DT_SIM);
    accumulateDose(whiteDose, whiteInner, DT_SIM);

    if (yolkCentre > peakYolk) {
      peakYolk = yolkCentre;
      peakYolkTime = t;
    }
    if (whiteInner > peakWhite) peakWhite = whiteInner;
    if (!pullRecorded && t >= cookTime_s) {
      yolkAtPull = yolkCentre;
      pullRecorded = true;
    }

    // Stop once the egg is past its peak and cooling, and the remaining dose
    // rate is a millionth of the peak rate - further integration cannot change
    // the answer. Because z ~ 4.65 K the rate collapses fast, so this typically
    // cuts an ice-bath simulation in half.
    const rate = Math.pow(10.0, (yolkCentre - TREF_YOLK_C) / Z_YOLK);
    if (rate > peakDoseRate) peakDoseRate = rate;
    if (t > cookTime_s && yolkCentre < prevYolk && rate < 1e-6 * peakDoseRate) break;
    prevYolk = yolkCentre;
  }

  return {
    cookTime_s: cookTime_s,
    peakYolk_C: peakYolk,
    peakYolkTime_s: peakYolkTime,
    yolkAtPull_C: yolkAtPull,
    yolkDose_min: yolkDose.minutes,
    whiteDose_min: whiteDose.minutes,
    peakWhite_C: peakWhite,
  };
}

/* ------------------------------------------------------------------ search */

const SOLVE_LO_S = 20.0;
const SOLVE_HI_S = 3600.0;
const SOLVE_TOL_S = 1.0;

/** Which dose a search is looking at. */
type Metric = (r: CookResult) => number;

function yolkOf(r: CookResult): number { return r.yolkDose_min; }
function whiteOf(r: CookResult): number { return r.whiteDose_min; }

/** Cook time in [lo, hi] at which `metric` first reaches `target`, assuming
 *  the metric is monotonic across that bracket. Returns the upper end of the
 *  final bracket: the shortest cook KNOWN to meet the target, to within
 *  SOLVE_TOL_S. The midpoint would be as accurate, but could sit a hair short
 *  of the target, and callers compare the dose at the answer against it. */
function bisectBetween(
  egg: Egg, setup: CookSetup, params: ModelParams,
  target: number, metric: Metric, lo_s: number, hi_s: number,
): number {
  let lo = lo_s;
  let hi = hi_s;
  while (hi - lo > SOLVE_TOL_S) {
    const mid = 0.5 * (lo + hi);
    if (metric(simulate(egg, setup, params, mid)) < target) lo = mid;
    else hi = mid;
  }
  return hi;
}

/** Bisect the whole search range. Both doses are monotonically increasing in
 *  cook time while the water is held at the boil (test 16a pins this), so
 *  bisection is exact there. See solveStanding for what happens when the heat
 *  goes off. */
function bisect(
  egg: Egg, setup: CookSetup, params: ModelParams, target: number, metric: Metric,
): number {
  return bisectBetween(egg, setup, params, target, metric, SOLVE_LO_S, SOLVE_HI_S);
}

export interface Solution {
  result: CookResult;
  /** False when the requested yolk doneness cannot be reached without leaving
   *  the white undercooked - the classic case being a rested egg, where
   *  carryover drives the yolk past soft no matter how short the cook. */
  reachable: boolean;
  /** Shortest cook that still sets the white. */
  minCookTime_s: number;
  /** Softest yolk doneness achievable with this setup, as a slider position. */
  softestLevel: number;
  /** Hardest yolk doneness achievable, as a slider position. 1 whenever the
   *  water stays at the boil; less than 1 when the heat is off and the pan
   *  runs out of heat before the yolk gets there. */
  hardestLevel: number;
  /** False when the white never sets at all - only possible with the heat off,
   *  where the water can fall past the white's own target while the egg is
   *  still in it. There is no cook time to offer in that case, only the
   *  furthest this pan goes. */
  whiteSets: boolean;
}

function solutionOf(
  result: CookResult, reachable: boolean, minCookTime_s: number,
  softestLevel: number, hardestLevel: number, whiteSets: boolean,
): Solution {
  return {
    result: result, reachable: reachable, minCookTime_s: minCookTime_s,
    softestLevel: softestLevel, hardestLevel: hardestLevel, whiteSets: whiteSets,
  };
}

/** Solve for the cook time that delivers the requested doneness. */
export function solveCookTime(
  egg: Egg, setup: CookSetup, params: ModelParams, doneness: Doneness,
): Solution {
  if (setup.afterBoil === 'off') return solveStanding(egg, setup, params, doneness);

  const minCook = bisect(egg, setup, params, doneness.whiteDose_min, whiteOf);
  const atMin = simulate(egg, setup, params, minCook);
  const softestLevel = sliderFromYolkDose(atMin.yolkDose_min);
  const whiteSets = atMin.whiteDose_min >= doneness.whiteDose_min;

  if (atMin.yolkDose_min >= doneness.yolkDose_min) {
    // Even the shortest white-setting cook overcooks the yolk past the target.
    return solutionOf(atMin, false, minCook, softestLevel, 1.0, whiteSets);
  }

  const cook = bisect(egg, setup, params, doneness.yolkDose_min, yolkOf);
  return solutionOf(
    simulate(egg, setup, params, cook), true, minCook, softestLevel, 1.0, whiteSets,
  );
}

/* --------------------------------------------------------------- standing */

/** Fraction of the best available dose that counts as "as far as this pan
 *  goes". The maximum sits on a plateau - the last per cent of the dose can
 *  take another quarter of an hour and change the yolk by a tenth of a degree -
 *  so the time worth printing is the start of that plateau, not its peak. */
const STANDING_KNEE = 0.99;

/** How long past the boil it is worth looking, with the heat off. The water is
 *  falling; half an hour after the burner dies there is nothing left to give,
 *  and a longer search only costs simulations. */
const STANDING_HORIZON_S = 1800.0;

/** Coarse step for the standing scan, seconds. Fine enough that the bracket it
 *  hands to the bisection is locally monotonic; coarse enough to keep the scan
 *  to a few dozen simulations. */
const SCAN_STEP_S = 30.0;

/** Both doses sampled at every scan step, from the shortest cook to the
 *  horizon. One pass answers every question the standing solver has. */
interface DoseCurve {
  times_s: number[];
  yolk: number[];
  white: number[];
}

function scanStanding(
  egg: Egg, setup: CookSetup, params: ModelParams, horizon_s: number,
): DoseCurve {
  const times: number[] = [];
  const yolk: number[] = [];
  const white: number[] = [];
  for (let t = SOLVE_LO_S; t <= horizon_s; t += SCAN_STEP_S) {
    const r = simulate(egg, setup, params, t);
    times.push(t);
    yolk.push(r.yolkDose_min);
    white.push(r.whiteDose_min);
  }
  return { times_s: times, yolk: yolk, white: white };
}

/** Index of the largest sample. The curve is never empty: the horizon is at
 *  least STANDING_HORIZON_S past a non-negative time to boil. */
function indexOfMax(values: number[]): number {
  let best = 0;
  for (let i = 1; i < values.length; i++) {
    if (values[i] > values[best]) best = i;
  }
  return best;
}

/**
 * First cook time at which a sampled dose reaches `target`, WITHOUT assuming
 * monotonicity - or -1 if it never does.
 *
 * Held at the boil, a longer cook always means more dose, and `bisect` is
 * exact. With the heat off that is false: pulling later means pulling from
 * cooler water, so the carryover that follows is smaller, and past a certain
 * point the total dose FALLS with a longer cook. Bisection on a non-monotonic
 * function does not merely lose accuracy - it lands anywhere, which showed up
 * as cook times jumping between 3 and 13 minutes for a 5 C change in room
 * temperature.
 *
 * So: walk the samples until the target is first met, then bisect inside that
 * one step, where the function is still rising.
 */
function firstCrossing(
  egg: Egg, setup: CookSetup, params: ModelParams,
  curve: DoseCurve, values: number[], target: number, metric: Metric,
): number {
  for (let i = 0; i < values.length; i++) {
    if (values[i] < target) continue;
    if (i === 0) return curve.times_s[0];
    return bisectBetween(
      egg, setup, params, target, metric, curve.times_s[i - 1], curve.times_s[i],
    );
  }
  return -1.0;
}

/**
 * The same question with the heat off, where neither dose is monotonic in cook
 * time and every shortcut above stops being valid.
 *
 * Scanning is unavoidable, but it is bounded: the water is falling, so past
 * roughly half an hour of standing nothing changes at all, and there is no
 * point looking further. The scan runs to the horizon regardless of where the
 * target is crossed, so `hardestLevel` is the true ceiling of this pan and
 * not merely "at least what was asked".
 */
function solveStanding(
  egg: Egg, setup: CookSetup, params: ModelParams, doneness: Doneness,
): Solution {
  const curve = scanStanding(egg, setup, params, setup.timeToBoil_s + STANDING_HORIZON_S);
  const whiteCook = firstCrossing(
    egg, setup, params, curve, curve.white, doneness.whiteDose_min, whiteOf,
  );
  const yolkCook = firstCrossing(
    egg, setup, params, curve, curve.yolk, doneness.yolkDose_min, yolkOf,
  );

  const peak = indexOfMax(curve.yolk);
  const maxDose = curve.yolk[peak];
  const maxAt = curve.times_s[peak];
  const hardestLevel = sliderFromYolkDose(maxDose);
  const whiteSets = whiteCook > 0;
  const minCook = whiteSets ? whiteCook : maxAt;
  const atMin = simulate(egg, setup, params, minCook);
  const softestLevel = sliderFromYolkDose(atMin.yolkDose_min);

  if (!whiteSets || yolkCook < 0) {
    // Either the water never gets the white where it needs to go, or it runs
    // out before the yolk does. Answer with the furthest this pan goes, rather
    // than with a time that does not deliver what was asked - but take the
    // START of the plateau, since waiting past it achieves nothing.
    const knee = firstCrossing(
      egg, setup, params, curve, curve.yolk, STANDING_KNEE * maxDose, yolkOf,
    );
    const at = knee > 0 ? knee : maxAt;
    return solutionOf(
      simulate(egg, setup, params, at), false, minCook, softestLevel, hardestLevel, whiteSets,
    );
  }

  if (atMin.yolkDose_min >= doneness.yolkDose_min) {
    return solutionOf(atMin, false, minCook, softestLevel, hardestLevel, whiteSets);
  }

  // Both constraints have to hold at the same pull, so take the later crossing.
  const cook = yolkCook > whiteCook ? yolkCook : whiteCook;
  return solutionOf(
    simulate(egg, setup, params, cook), true, minCook, softestLevel, hardestLevel, whiteSets,
  );
}
